### DISCUSSION TITLE
Modifying and saving a workbook

### DISCUSSION THREAD

**Discussion by don-odyssey** [PHPOffice/PhpSpreadsheet#3621] [Q&A]
Hi. I've figured out how to load a workbook and write to cells in a spreadsheet. I would now like to save it as a ew file in a different directory than the file loaded. I can't figure out how to do this?

**Comment by MarkBaker** [comment]
Instantiate a Writer, passing it the Spreadsheet as an argument, then use the save method, passing it a folder/filename

**Comment by oleibman** [comment]
```php
$directory = 'directoryname';
$outputFile = 'filename.xlsx';
$writer = new \PhpOffice\PhpSpreadsheet\Writer\Xlsx($spreadsheet);
$writer->save("$directoryname/$outputFile");
```

**Comment by don-odyssey** [comment]
<?php
use PhpOffice\PhpSpreadsheet\IOFactory;

$importFile = '/work/dosc/myworkbook.xlsx';
$exportFile = '/work/dosc/sent/myworkbook_revised.xlsx';

$reader = IOFactory::createReader("Xlsx");
$workbook = $reader->load($importFile);

$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');

$outboundWorkbook = new \PhpOffice\PhpSpreadsheet\Spreadsheet();
$writer = IOFactory::createWriter($outboundWorkbook, "Xlsx");
$writer->save($exportFile);
?>

**Comment by don-odyssey** [comment]
Sorry but I don't know how to format it properly so that is shows on separate lines

**Comment by don-odyssey** [comment]
It saves as a blank workbook. Here is my code

 

<?php

use PhpOffice\PhpSpreadsheet\IOFactory;

 

$importFile = '/work/dosc/myworkbook.xlsx';

$exportFile = '/work/dosc/sent/myworkbook_revised.xlsx';

 

$reader = IOFactory::createReader("Xlsx");

$workbook = $reader->load($importFile);

 

$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');

 

$outboundWorkbook = new \PhpOffice\PhpSpreadsheet\Spreadsheet();

$writer = IOFactory::createWriter($outboundWorkbook, "Xlsx");

$writer->save($exportFile);

?>

**Comment by don-odyssey** [comment]
It saves as a blank workbook. Here is my code

 

<?php

use PhpOffice\PhpSpreadsheet\IOFactory;

 

$importFile = '/work/dosc/myworkbook.xlsx';

$exportFile = '/work/dosc/sent/myworkbook_revised.xlsx';

 

$reader = IOFactory::createReader("Xlsx");

$workbook = $reader->load($importFile);

 

$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');

 

$outboundWorkbook = new \PhpOffice\PhpSpreadsheet\Spreadsheet();

$writer = IOFactory::createWriter($outboundWorkbook, "Xlsx");

$writer->save($exportFile);

?>

**Comment by MarkBaker** [comment]
You're instantiating a new Spreadsheet object and passing that as the argument to the Writer.
Pass in the Spreadsheet object that you loaded.

**Comment by don-odyssey** [comment]
That’s the confusing part. If I do that, it will overwrite the original file instead of creating a new file.

**Comment by don-odyssey** [comment]
I see something about cloning a spreadsheet. Do I need to do that first before writing? How to write to a different location?

**Comment by oleibman** [comment]
```php
<?php

use PhpOffice\PhpSpreadsheet\IOFactory;

$importFile = '/work/dosc/myworkbook.xlsx';
$exportFile = '/work/dosc/sent/myworkbook_revised.xlsx';

$reader = IOFactory::createReader("Xlsx");
$workbook = $reader->load($importFile);

$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');

// the next 2 statement are where you are going wrong
//$outboundWorkbook = new \PhpOffice\PhpSpreadsheet\Spreadsheet();
//$writer = IOFactory::createWriter($outboundWorkbook, "Xlsx");

// use the following single statement in place of those 2
$writer = IOFactory::createWriter($workbook, "Xlsx");

$writer->save($exportFile);
```
To format your code nicely, use a line with 3 back-ticks followed by the 3 characters php.
Then include your code.
Then finish the operation by using a line with 3 back-ticks.

**Comment by don-odyssey** [comment]
Ok thanks.I’ll repost my question with the code as a new post since I haven’t yet received a solution.Sent from my iPhoneOn Jun 19, 2023, at 6:21 PM, oleibman ***@***.***> wrote:﻿
<?php

use PhpOffice\PhpSpreadsheet\IOFactory;

$importFile = '/work/dosc/myworkbook.xlsx';
$exportFile = '/work/dosc/sent/myworkbook_revised.xlsx';

$reader = IOFactory::createReader("Xlsx");
$workbook = $reader->load($importFile);

$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');

// the next 2 statement are where you are going wrong
//$outboundWorkbook = new \PhpOffice\PhpSpreadsheet\Spreadsheet();
//$writer = IOFactory::createWriter($outboundWorkbook, "Xlsx");

// use the following single statement in place of those 2
$writer = IOFactory::createWriter($workbook, "Xlsx");

$writer->save($exportFile);
To format your code nicely, use a line with 3 back-ticks followed by the 3 characters php.
Then include your code.
Then finish the operation by using a line with 3 back-ticks.

**Comment by don-odyssey** [comment]
Oh O noticed you did respond with the fix, thank you! I will try tomorrow Sent from my iPhoneOn Jun 19, 2023, at 6:21 PM, oleibman ***@***.***> wrote:﻿
<?php

use PhpOffice\PhpSpreadsheet\IOFactory;

$importFile = '/work/dosc/myworkbook.xlsx';
$exportFile = '/work/dosc/sent/myworkbook_revised.xlsx';

$reader = IOFactory::createReader("Xlsx");
$workbook = $reader->load($importFile);

$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');

// the next 2 statement are where you are going wrong
//$outboundWorkbook = new \PhpOffice\PhpSpreadsheet\Spreadsheet();
//$writer = IOFactory::createWriter($outboundWorkbook, "Xlsx");

// use the following single statement in place of those 2
$writer = IOFactory::createWriter($workbook, "Xlsx");

$writer->save($exportFile);
To format your code nicely, use a line with 3 back-ticks followed by the 3 characters php.
Then include your code.
Then finish the operation by using a line with 3 back-ticks.

**Comment by don-odyssey** [comment]
Hi,

After making your modification, it is still throwing an error:

Fatal error: Uncaught PhpOffice\PhpSpreadsheet\Calculation\Exception: Table Table26 for Structured Reference cannot be located in /home/httpd/www/my-directory/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Calculation/Engine/Operands/StructuredReference.php:150

Code as follows
==========
```
<?php
use PhpOffice\PhpSpreadsheet\IOFactory;

$importTariffFile = '/work/dosc/myworkbook.xlsx';
$exportTariffFile = '/work/dosc/sent/myworkbook_revised.xlsx';

$reader = IOFactory::createReader("Xlsx");
$workbook = $reader->load($importTariffFile);

$workbook->setActiveSheetIndexByName('mysheet');
$workbook->getActiveSheet()->setCellValue('B7', 'Hello World');
    
$writer = IOFactory::createWriter($workbook, "Xlsx");
$writer->save($exportTariffFile);
?>

**Comment by oleibman** [comment]
I would need to see your workbook in order to attempt to debug this problem. I do not know much about Structured References. You can probably work around the problem by adding the following to your code before you save the file:
```php
$writer->setPreCalculateFormulas(false);
```
